c_glib/arrow-glib/basic-array.cpp

```diff
@@ -2105,6 +2110,54 @@ garrow_time64_array_get_values(GArrowTime64Array *array,
   return reinterpret_cast<const gint64 *>(values);
 }
 
+
+G_DEFINE_TYPE(GArrowFixedSizeBinaryArray,
+              garrow_fixed_size_binary_array,
+              GARROW_TYPE_PRIMITIVE_ARRAY)
+static void
+garrow_fixed_size_binary_array_init(GArrowFixedSizeBinaryArray *object)
+{
+}
+
+static void
+garrow_fixed_size_binary_array_class_init(GArrowFixedSizeBinaryArrayClass *klass)
+{
+}
+
+
+G_DEFINE_TYPE(GArrowDecimal128Array,
+              garrow_decimal128_array,
+              GARROW_TYPE_FIXED_SIZE_BINARY_ARRAY)
+static void
+garrow_decimal128_array_init(GArrowDecimal128Array *object)
+{
+}
+
+static void
+garrow_decimal128_array_class_init(GArrowDecimal128ArrayClass *klass)
+{
+}
+
+/**
+ * garrow_decimal128_array_get_value:
+ * @array: A #GArrowDecimal128Array.
+ * @i: The index of the target value.
+ *
+ * Returns: The i-th value.
+ *
+ * Since: 0.10.0
+ */
+gchar *
```

Comment:
How about returning `GArrowDecimal128` instead of `gchar *` as a follow-up pull request?
If we return `GArrowDecimal128`, we can convert the value to `gchar *`, `gint64` and so on.

Reply:
@kou 
Please confirm my understanding.

```
GArrowDecimal128 *
garrow_decimal128_array_get_value(GArrowDecimal128Array *array,
                                  gint64 i)
{
  auto arrow_array = garrow_array_get_raw(GARROW_ARRAY(array));
  auto arrow_decimal128_array =
    static_cast<arrow::Decimal128Array *>(arrow_array.get());
  auto value = arrow_decimal128_array->FormatValue(i);
  auto arrow_decimal = std::make_shared<arrow::Decimal128>(value);
  return garrow_decimal128_new_raw(&arrow_decimal);
}
```

```
def test_value
    data_type = Arrow::DecimalDataType.new(8,2)
    builder = Arrow::Decimal128ArrayBuilder.new(data_type)
    decimal = Arrow::Decimal128.new("23423445")
    builder.append(decimal)
    array = builder.finish
    assert_equal("23423445", array.get_value(0).to_s)
  end
```

In the code above, GArrowDecimal128 does not include scale.



Does GArrowDecimal128 need to return including scale?
I don't know how to return GArrowDecimal128 including scale.

Reply:
The scale information has `GArrowDecimalDataType` not `GArrowDecimal128`. We can get `GArrowDecimalDataType` from `array.value_data_type`.
We don't have `garrow_decimal_data_type_get_scale()` (and `_precision()`) for now. When we create them, we can write like the following:

```diff
diff --git a/c_glib/arrow-glib/basic-array.cpp b/c_glib/arrow-glib/basic-array.cpp
index 92230561..00e1746c 100644
--- a/c_glib/arrow-glib/basic-array.cpp
+++ b/c_glib/arrow-glib/basic-array.cpp
@@ -2143,19 +2143,20 @@ garrow_decimal128_array_class_init(GArrowDecimal128ArrayClass *klass)
  * @array: A #GArrowDecimal128Array.
  * @i: The index of the target value.
  *
- * Returns: The i-th value.
+ * Returns: (transfer full): The i-th value.
  *
  * Since: 0.10.0
  */
-gchar *
+GArrowDecimal128 *
 garrow_decimal128_array_get_value(GArrowDecimal128Array *array,
                                   gint64 i)
 {
   auto arrow_array = garrow_array_get_raw(GARROW_ARRAY(array));
   auto arrow_decimal128_array =
     static_cast<arrow::Decimal128Array *>(arrow_array.get());
-  auto value = arrow_decimal128_array->FormatValue(i);
-  return g_strndup(value.data(), value.size());
+  auto arrow_decimal =
+    std::make_shared<arrow::Decimal128>(arrow_decimal128_array->GetValue(i));
+  return garrow_decimal128_new_raw(&arrow_decimal);
 }
 
 G_END_DECLS
diff --git a/c_glib/arrow-glib/basic-array.h b/c_glib/arrow-glib/basic-array.h
index 957e938c..ac20fec7 100644
--- a/c_glib/arrow-glib/basic-array.h
+++ b/c_glib/arrow-glib/basic-array.h
@@ -654,7 +654,7 @@ struct _GArrowDecimal128ArrayClass
   GArrowFixedSizeBinaryArrayClass parent_class;
 };
 
-gchar *garrow_decimal128_array_get_value(GArrowDecimal128Array *array,
-                                         gint64 i);
+GArrowDecimal128 *garrow_decimal128_array_get_value(GArrowDecimal128Array *array,
+                                                    gint64 i);
 
 G_END_DECLS
diff --git a/c_glib/test/test-decimal-array.rb b/c_glib/test/test-decimal-array.rb
index 303cc071..f75b12c9 100644
--- a/c_glib/test/test-decimal-array.rb
+++ b/c_glib/test/test-decimal-array.rb
@@ -22,6 +22,7 @@ class TestDecimalArray < Test::Unit::TestCase
     decimal = Arrow::Decimal128.new("23423445")
     builder.append(decimal)
     array = builder.finish
-    assert_equal("234234.45", array.get_value(0))
+    assert_equal("234234.45",
+                 array.get_value(0).to_string_scale(array.value_data_type.scale))
   end
 end
```

Reply:
Thanks.
I understand what you meant.
I'll create follow-up pull requests to resolve your comments.

Reply:
Good!

i thought about API. It's better that the current `get_value` function is remained and renamed to `format_value`. Because it'll be useful for some cases. Here is a patch to show this concept:

```diff
diff --git a/c_glib/arrow-glib/basic-array.h b/c_glib/arrow-glib/basic-array.h
index 957e938c..23891698 100644
--- a/c_glib/arrow-glib/basic-array.h
+++ b/c_glib/arrow-glib/basic-array.h
@@ -654,7 +654,7 @@ struct _GArrowDecimal128ArrayClass
   GArrowFixedSizeBinaryArrayClass parent_class;
 };
 
-gchar *garrow_decimal128_array_get_value(GArrowDecimal128Array *array,
-                                         gint64 i);
+gchar *garrow_decimal128_array_format_value(GArrowDecimal128Array *array,
+                                            gint64 i);
 
 G_END_DECLS
diff --git a/c_glib/arrow-glib/basic-array.cpp b/c_glib/arrow-glib/basic-array.cpp
index 92230561..098dc5e6 100644
--- a/c_glib/arrow-glib/basic-array.cpp
+++ b/c_glib/arrow-glib/basic-array.cpp
@@ -2145,11 +2145,14 @@ garrow_decimal128_array_class_init(GArrowDecimal128ArrayClass *klass)
  *
  * Returns: The i-th value.
  *
+ * The returned string should be freed with g_free() when no longer
+ * needed.
+ *
  * Since: 0.10.0
  */
 gchar *
-garrow_decimal128_array_get_value(GArrowDecimal128Array *array,
-                                  gint64 i)
+garrow_decimal128_array_format_value(GArrowDecimal128Array *array,
+                                     gint64 i)
 {
   auto arrow_array = garrow_array_get_raw(GARROW_ARRAY(array));
   auto arrow_decimal128_array =
```

Reply:
Thanks!
That makes sense. I'll rename it.